File: probability-example-main/probability-example-PythonVersion/python/GameRunner.py
import Coin
import Die
import Deck

class GameRunner:
    def __init__(self):
        self.coin = Coin.Coin()
        self.die = Die.Die()
        self.deck = Deck.Deck()
        self.coin_points = 1;
        self.dice_points = 2;
        self.card_points = 5;

    def play_coin_game(self, points):
        play_again = 'Y'
        while play_again != 'N':
            first_flip = self.coin.flip()
            print("\nFlipping coin!\n"
                    f"{self.coin.display()}\n"
                    "Do you think the next flip will be the same? Y or N: ");
            choice = GameRunner.get_y_or_n()
            print("\nFlipping again!")
            second_flip = self.coin.flip()
            print(f"{self.coin.display()}")
            win = False
            if (choice == 'Y' and second_flip == first_flip):
                win = True
            elif (choice == 'N' and second_flip != first_flip):
                win = True
            
            if win:
                points += self.coin_points
                print(f"You win! You now have {points} points!")
            else:
                points -= self.coin_points
                print(f"You lose! You now have {points} points!")

            if points < self.coin_points:
                play_again = 'N'
                print("Sorry, you don't have enough points to keep playing.")
            else:
                print("\nWould you like to play again? Y or N: ")
                play_again = GameRunner.get_y_or_n()

        return points

    def play_dice_game(self, points):
        play_again = 'Y'
        while play_again != 'N':
            first_roll = self.die.roll()
            print(f"{self.die.display()}\n"
                   "Do you think the next roll will be Higher (>), Lower (<), or the same (=) ?\n"
                   "Enter >, <, or = ?")
            choice = GameRunner.get_compare()
            print("\nRolling again!")
            second_roll = self.die.roll()
            print(f"{self.die.display()}")
            win = False
            if (choice == '>' and second_roll > first_roll):
                win = True
            elif (choice == '=' and second_roll == first_roll):
                win = True
            elif (choice == '<' and second_roll < first_roll):
                win = True
            
            if win:
                points += self.dice_points
                print(f"You win! You now have {points} points!")
            else:
                points -= self.dice_points
                print(f"You lose! You now have {points} points!")

            if points < self.dice_points:
                play_again = 'N'
                print("Sorry, you don't have enough points to keep playing.")
            else:
                print("\nWould you like to play again? Y or N: ")
                play_again = GameRunner.get_y_or_n()

        return points

    def play_card_game(self, points):
        play_again = 'Y'
        self.deck.shuffle_all()
        while play_again != 'N':
            first_card = self.deck.draw_and_discard()
            second_card = self.deck.draw_and_discard()
            print("\nDealing card\n"
                    f"{first_card.display_reverse(' of ')}\n"
                    # f"{second_card.display_reverse(' of ')}\n" # Debugging
                    "Do you think the next card will be higher (>), lower (<), or the same (=) ?\n"
                    "Enter >, <, or = : ")
            choice = GameRunner.get_compare()
            print(f"Next Card!\n{second_card.display_reverse(' of ')}")
            win = False
            if (choice == '<' and Deck.Deck.compare(second_card, first_card) == Deck.Comparator.LESS):
                win = True
            elif (choice == '=' and Deck.Deck.compare(second_card, first_card) == Deck.Comparator.EQUAL):
                win = True
            elif (choice == '>' and Deck.Deck.compare(second_card, first_card) == Deck.Comparator.GREATER):
                win = True

            if win:
                points += self.card_points
                print(f"You win!  You now have {points} points!")
            else:
                points -= self.card_points
                print(f"You lose!  You now have {points} pionts!")

            if points < self.card_points:
                play_again = 'N'
                print("Sorry. You don't have enough points to keep playing.")
            else:
                print("Would you like to play again?  Y or N: ")
                play_again = GameRunner.get_y_or_n()

        return points

    def get_compare():
        choice = input()
        while choice not in ('>', '=', '<'):
            print("Please enter >, <, or =: ")
        return choice

    def get_y_or_n():
        choice = input().upper()
        while choice not in ('Y', 'N'):
            print("Please enter Y or N: ")
            choice = input().upper()
        return choice

File: probability-example-main/probability-example-PythonVersion/csharp/GameRunner.cs
/*This is the template for a GameRunner object. It is responsible for running
all of the games.*/

public class GameRunner{

    private Coin _coin = new Coin();
    public int coinPoints = 1; //how many points the coin game is worth
    private Die _die = new Die();
    public int dicePoints = 2; //how many points the dice game is worth
    public Deck _deck = new Deck();
    public int cardPoints = 5;

    //Reads input line and verfies a Y or N
    //Returns Y or N    
    private string getYN(){
        string choice = Console.ReadLine().ToUpper();
        while (choice != "Y" && choice != "N"){
            Console.Write("Please enter Y or N: ");
            choice = Console.ReadLine().ToUpper();
        } 
        return choice;
    }
    //Reads input line and verfies a >, < or =
    //Returns the symbol   
    private string getCompare(){
        string choice = Console.ReadLine();
        while (choice != ">" && choice != "<" && choice != "="){
            Console.Write("Please enter >, < or =: ");
            choice = Console.ReadLine();
        } 
        return choice;
    }


    //Flips a coin. 
    //Player guess if next is same or different
    //Flips a coin.
    //Awards or sutracts points if player was correct or not.
    //Allows repeated playing, unles player does not have enough points.
    public int CoinGame(int points){
        string playAgain = "Y";
        while (playAgain != "N"){
            Console.WriteLine("\nFlipping coin!");
            bool firstFlip = _coin.Flip();
            Console.WriteLine(_coin.Display());
            Console.Write("Do you think the next coin flip will be the same? Y or N: ");
            string bet = getYN();
            Console.WriteLine("\nFlipping again!");
            bool secondFlip = _coin.Flip();
            Console.WriteLine(_coin.Display());
            if(bet == "Y"){
                if(firstFlip == secondFlip){
                    points += coinPoints;
                    Console.WriteLine($"You win! You now have {points} points!");
                }
                else{
                    points -= coinPoints;
                    Console.WriteLine($"You lose! You now have {points} points!");
                }
            }
            else{
                if(firstFlip != secondFlip){
                    points += coinPoints;
                    Console.WriteLine($"You win! You now have {points} points!");
                }
                else{
                    points -= coinPoints;
                    Console.WriteLine($"You lose! You now have {points} points!");
                }
            }

            if(points < coinPoints){
                playAgain = "N";
                Console.WriteLine("Sorry. You don't have enough points to keep playing.");
            }
            else{
                Console.Write("\nWould you like to play again? Y or N: ");
                playAgain = getYN();
            }
        }
        return points;
    }//end CoinGame

    //Roll die.
    //Allows user to guess if new roll higher or lower.
    //Roll die.
    //Awards or subtracts points if user is correct or not.
    //Allows for repetition if player ahs enough points.
    //TODO: variable points depending on probability of correct guess?
    public int DiceGame (int points){
        string playAgain = "Y";
        while (playAgain != "N"){
            Console.WriteLine("\nRolling die!");
            int firstRoll = _die.Roll();
            Console.WriteLine(_die.Display());
            Console.WriteLine("Do you think the next roll will be Higher(>), Lower(<) or the Same(=)?");
            Console.Write("Enter >, < or =? ");
            string bet = getCompare();
            Console.WriteLine("\nRolling again!");
            int secondRoll = _die.Roll();
            Console.WriteLine(_die.Display());
            //Check if the user won
            if(bet == ">"){
                if(secondRoll > firstRoll){
                    points += dicePoints;
                    Console.WriteLine($"You win! You now have {points} points!");
                }
                else{
                    points -= dicePoints;
                    Console.WriteLine($"You lose! You now have {points} points!");
                }
            }
            else if (bet == "<"){
                if(secondRoll < firstRoll){
                    points += dicePoints;
                    Console.WriteLine($"You win! You now have {points} points!");
                }
                else{
                    points -= dicePoints;
                    Console.WriteLine($"You lose! You now have {points} points!");
                }
            }
            else{
                if(secondRoll == firstRoll){
                    points += dicePoints;
                    Console.WriteLine($"You win! You now have {points} points!");
                }
                else{
                    points -= dicePoints;
                    Console.WriteLine($"You lose! You now have {points} points!");
                }
            }
            //Check if playing again
            if(points < dicePoints){
                playAgain = "N";
                Console.WriteLine("Sorry. You don't have enough points to keep playing.");
            }
            else{
                Console.Write("\nWould you like to play again? Y or N: ");
                playAgain = getYN();
            }
        }
        return points;
    }//end DiceGame
    
    public int CardGame (int points){
        string playAgain = "Y";
        _deck.ShuffleAll();
        while (playAgain != "N"){
            Console.WriteLine("\nDealing card");
            Card firstCard = _deck.DrawAndDiscard();
            Console.WriteLine(firstCard.DisplayReverse(" of "));
            Card secondCard = _deck.DrawAndDiscard();
            Console.WriteLine("Do you think the next card will be Higher(>), Lower(<) or the Same(=)?");
            Console.Write("Enter >, < or =? ");
            string bet = getCompare();
            Console.WriteLine("\nNext Card!");
            Console.WriteLine(secondCard.DisplayReverse(" of "));
            //Check if the user won
            if(bet == ">"){
                if(Deck.Compare(secondCard,firstCard) == Comparator.Greater){
                    points += cardPoints;
                    Console.WriteLine($"You win! You now have {points} points!");
                }
                else{
                    points -= cardPoints;
                    Console.WriteLine($"You lose! You now have {points} points!");
                }
            }
            else if (bet == "<"){
                if(Deck.Compare(secondCard,firstCard) == Comparator.Less){
                    points += cardPoints;
                    Console.WriteLine($"You win! You now have {points} points!");
                }
                else{
                    points -= cardPoints;
                    Console.WriteLine($"You lose! You now have {points} points!");
                }
            }
            else{
                if(Deck.Compare(secondCard,firstCard) == Comparator.Equal){
                    points += cardPoints;
                    Console.WriteLine($"You win! You now have {points} points!");
                }
                else{
                    points -= cardPoints;
                    Console.WriteLine($"You lose! You now have {points} points!");
                }
            }
            //Check if playing again
            if(points < cardPoints){
                playAgain = "N";
                Console.WriteLine("Sorry. You don't have enough points to keep playing.");
            }
            else{
                Console.Write("\nWould you like to play again? Y or N: ");
                playAgain = getYN();
            }
        }
        return points;
    }//end CardGame
}
